ARCHITECTURE.md

```diff
@@ -180,16 +180,80 @@ Issues with this approach:
-
----
-
 Also, Yo-Yo (Bel) has some issues:
 
 - Local network requests from `<img src="">` are made on each state update. Not a big deal, but annoying and does not happing with vdom solutions, like Preact.
 - Webcam currently flashes when state is updated (morphdom issue?).
 - When using template strings it might be harder to re-use UI components in React, where JSX is a standard.
 
 We are thinking about trying Preact for more stability and React compatibility, but not sure if it’s worth it.
+
+
+## Harry's Proposal
+---
+My proposal for Uppy's architecture and API address two use cases separately:
+
+1. Uppy when its used without React and state is managed internally
+
+2. Uppy when it's used with React, and the user manages their own state (or we provide a React container component that takes care of it for them.)  I'll refer to them as the "vanilla/yo-yo" and "React" use cases, respectively.
+
+### Vanilla (using yo-yo) with managed state
+
+So we have a problem. When we decouple rendering from Uppy, and a user is able to make their own custom UI, they'll need to be able to call on plugin methods to do things, like fetch a list of files from Google Drive, for example.  In our current setup, the core creates and holds an instance of every plugin, and outside UIs don't have direct access to these plugin instances.  To get around this, one idea we was to expose a bus/event emitter, and they could emit something like `bus.emit('google-drive:list-files', { directory: 'root' })` when they wanted to fetch a list of files.
+
+I'm not a big fan of this.  The user becomes responsible for knowing the names of every single plugin event, which is prone to typos, and, in my opinion, it's a better pattern to be able to call functions directly to make these things happen vs. emitting events.
+
+My current proposal is to use Redux, or at least take a Redux-like approach, and have each plugin define a reducer and some action creators as an interface.  The core combines each reducer into a single one and holds it, along with a Redux store.
+
+Uppy would then pass down the Redux state, some core methods, and the dispatcher as "props" to the component, and the component would decide what to do with it all.
+
+The user code would look something like:
+
+```js
+import Uppy, { Remote, Uploader } from 'uppy'
+import DriveBrowser from 'uppy-ui'
+
+const uppy = new Uppy(someOptions)
+uppy.use(Remote, { source: 'drive' })
+uppy.use(Uploader)
+
+uppy.render(DriveBrowser, document.getElementById('root'))
+```
+
+The DriveBrowser component might look like:
+
+```js
+import { Remote } from 'uppy'
+const googleDrive = new Remote({ source: 'drive' })
+const { actions } = googleDrive
+
+const DriveBrowser = (props) => {
+  return html`
+    <div>
+      <button onclick=${props.dispatch(actions.list('root'))}>Get list of files</button>
+      /* ... */
+    </div>
+  `
+}
+```
+
```

Comment:
> props.dispatch(actions.list('root'))

Does this mean it’s including the function, `action.list('root')` in the dispatch?